### ISSUE TITLE
"Java language server exited unexpectedly" Error on macOS Monterey

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#1120]
@JustinProminic has been complaining about some problems on macOS Monterey.  I got access to a test machine with macOS 12.6.1 (Justin was on 12.5.1), and I see this behavior in Moonshine 3.3.3:
1. Close all other projects
2. Create a new Java Gradle application.  BUG:  language server reports error below.
3. Reopen the project.  BUG:  Same language server error
4. Project > Run Gradle Command.  This works properly, so it is not a problem with the JDK or with permissions for the JDK.
5. Reopen the project.  BUG:  Same language server error
6. Open the project settings and change the Gradle action to "eclipse".  I was hoping to reset the language server configuration with this.
7. Project > Run Gradle Command.  This runs with no errors.
8. Reopen the project.  BUG:  Same language server error

    : Java language server exited unexpectedly. Close the <project> project and re-open it to enable code intelligence.

If I run the same test on an older version of macOS (10.15.7 Catalina), the language server starts with no errors.

**Comment by JoelProminic** [comment]
My Monterey mac is pretty fresh, so I checked the Security & Privacy settings.  I didn't see any obviously disabled permissions, though.

I tried [giving Full Disk Access to Moonshine](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/980) and also to the JDK 11 `java` executable, but this didn't make a difference.

**Comment by joshtynjala** [comment]
I gave it a try, but I was unable to reproduce this issue. Creating a new Java Gradle project worked successfully on my machine right away at step 2. 

macOS 13 (Ventura)
Java 11.0.16.1

**Comment by JoelProminic** [comment]
That is unfortunate.  Do you have any ideas for other tests I can try to debug this?   Do you see any extra permissions granted to Java or Moonshine in Security & Privacy?

**Comment by Igazine** [comment]
I don't have the error running Moonshine in ADL, but the error is present in 3.3.3 and 3.3.4 dev.

macOS Monterey 12.6
OpenJDK 19

**Comment by joshtynjala** [comment]
> I don't have the error running Moonshine in ADL, but the error is present in 3.3.3 and 3.3.4 dev.

I also ran Moonshine in ADL.

**Comment by JoelProminic** [comment]
Hmm, do we support OpenJDK 19?   I'm on 11.0.10, the version installed by Moonshine SDK Installer.

I get the same error for a Java Maven project, but the language server works fine for a Flex Desktop or Haxe OpenFL project.

**Comment by Igazine** [comment]
Sorry, OpenJDK v 11.0.10

**Comment by joshtynjala** [comment]
It's worth mentioning that Moonshine 3.3.3 upgraded to a newer eclipse.jdt.ls language server. It was v1.9.0 in 3.3.2 and now it's v1.15.0 as of 3.3.3:

Relevant commit: a4cdb6e421eee3bf7d6d40b1bf7325d6749d845c

**Comment by joshtynjala** [comment]
Hmmm... according to the changelog, JDK 17 is required starting with eclipse.jdt.ls v1.13.0. That's interesting. I would normally say that explains why it crashes. However, surely that should also cause it to crash for me and @Igazine when running in ADL. Unless, somehow it is using a newer JDK version, even though we have JDK 11 configured in Moonshine.

**Comment by Igazine** [comment]
I can reproduce the error after purging the Application Support directory and restarting Moonshine in ADL.

@joshtynjala I noticed earlier that Moonshine copies the jdt-language-server files into its own Application Support directory, and once the files exist, they won't get overwritten next time there's an update, so the old version is still being used. Try cleaning up the directory and try ADL again.

**Comment by Igazine** [comment]
Important: obviously DO NOT purge the entire Application Support directory, only Moonshine's subdirectory. Sorry for the wording

**Comment by joshtynjala** [comment]
Moonshine is supposed to replace the eclipse.jdt.ls files copied to _Application Support_ when there's an updated version.

@Igazine Is it possible that your changes to _JavaLanguageServerManager.as_ has broken this behavior?

**Comment by Igazine** [comment]
I think you're right, the issue [resides in the wrapper's build config](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/d7cd84db22b39f5bf926e69304bde0ece6f185f1/language-server-wrappers/moonshine-jdt/build.xml#L58), give me a few minutes to look into it.

**Comment by joshtynjala** [comment]
@Igazine Here's part of the issue:

The value of `LANGUAGE_SERVER_JAR_FILE_NAME_PREFIX` was changed in commit c7aa79c723688ae46f3d86547b6e63de631a2be4. `JavaLanguageServerManager` searches for a file with that prefix, but the rest of the file name changes between versions of eclipse.jdt.ls. `JavaLanguageServerManager` was using the different file name to determine if it needed to update the files it copies into _Application Support_. With this change, it thinks that nothing has changed between versions because the file name is the same.

**Comment by Igazine** [comment]
I've reverted that change, added a new constant for the wrapper, but the original issue still occurs.

**Comment by Igazine** [comment]
Here's the Java log:

```
!SESSION 2022-11-02 00:22:09.475 -----------------------------------------------
eclipse.buildId=unknown
java.version=11.0.10
java.vendor=AdoptOpenJDK
BootLoader constants: OS=macosx, ARCH=x86_64, WS=cocoa, NL=en_HU
Command-line arguments:  -data /Users/tamas/Library/Application Support/idToBeReplacedByANT/Local Store/java/workspaces/a72a1f84f2a047d08c84079a5078df9821603c1afe41f875952197a118058212

!ENTRY org.eclipse.m2e.apt.core 4 0 2022-11-02 00:22:10.126
!MESSAGE FrameworkEvent ERROR
!STACK 0
org.osgi.framework.BundleException: Could not resolve module: org.eclipse.m2e.apt.core [55]
  Unresolved requirement: Require-Capability: osgi.ee; filter:="(&(osgi.ee=JavaSE)(version=17))"
  Unresolved requirement: Require-Bundle: org.eclipse.m2e.jdt; bundle-version="[2.0.0,3.0.0)"
    -> Bundle-SymbolicName: org.eclipse.m2e.jdt; bundle-version="2.0.0.20220706-1958"; singleton:="true"
       org.eclipse.m2e.jdt [57]
         Unresolved requirement: Require-Bundle: org.eclipse.m2e.core; bundle-version="[2.0.0,3.0.0)"
           -> Bundle-SymbolicName: org.eclipse.m2e.core; bundle-version="2.0.0.20220711-1014"; singleton:="true"
              org.eclipse.m2e.core [56]
                Unresolved requirement: Require-Capability: osgi.ee; filter:="(&(osgi.ee=JavaSE)(version=17))"
  Unresolved requirement: Require-Bundle: org.eclipse.m2e.core; bundle-version="[2.0.0,3.0.0)"
    -> Bundle-SymbolicName: org.eclipse.m2e.core; bundle-version="2.0.0.20220711-1014"; singleton:="true"

	at org.eclipse.osgi.container.Module.start(Module.java:463)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel$2.run(ModuleContainer.java:1847)
	at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor$1$1.execute(EquinoxContainerAdaptor.java:136)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1840)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1781)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1745)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1667)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:345)


```

**Comment by rat-moonshine** [comment]
It looks like guys already aware to the issue. I'm just subscribing because I'm having same trouble on Monterey starting lang-server against Java-Gradle.

`: Java language server exited unexpectedly. Close the HostManager2022OnlyExternalDns project and re-open it to enable code intelligence.`

**Comment by rat-moonshine** [comment]
It looks like guys already aware to the issue. I'm just subscribing because I'm having same trouble on Monterey starting lang-server against Java-Gradle.

`: Java language server exited unexpectedly. Close the $projectName project and re-open it to enable code intelligence.`

**Comment by Igazine** [comment]
My observations

- Changing `LANGUAGE_SERVER_JAR_FILE_NAME_PREFIX` to directly start the Eclipse plugin instead of the wrapper did not help, the problem is still present
- Forcing my OpenJDK 19 installation path manually in the code (overriding Moonshine's installed Java path) fixed the issue, language server started without errors
- However Gradle crashed after I created a new Gradle project
- After restarting Moonshine with the previously opened Gradle project, the previous crash did not occur, everything worked fine
- Creating a new Maven project also started the language server properly, no errors were present

So my assumption is that the new version of the Eclipse plugin does not work well with Java 11.

`Unresolved requirement: Require-Capability: osgi.ee; filter:="(&(osgi.ee=JavaSE)(version=17))"`

**Comment by JoelProminic** [comment]
I don't remember if we have a specific requirement for JDK 11, but we still need to support JDK 8 for the Java Domino and NSFODP projects.  The Domino API has a hard dependency on JDK 8 for now.

That said, I think we were using JDK 11 for the JDK 8 language server already, and ignoring the legacy errors.  We'll need to make sure that our new solution can handle this case.

**Comment by joshtynjala** [comment]
> but the original issue still occurs.

This is expected, since eclipse.jdt.ls seems to require JDK 17 now, according to its changelog. Now that our eclipse.jdt.ls update code should be working properly, I expect to be able to reproduce the issue, and I can switch to JDK 17 as the minimum required version for Java projects. I don't think it's a good idea to stay on an older eclipse.jdt.ls version, but I'd be willing to do it temporarily (reverting to v1.12.0, which should be the newest version that still supports JDK 11).

Assuming that we move forward with the JDK 17 requirement, I don't think that will be difficult. Since we already have logic to run eclipse.jdt.ls with JDK 11 while targeting JDK 8, it should be straightforward to run eclipse.jdt.ls with JDK 17 while targeting JDK 8. We would only run into issues if eclipse.jdt.ls somehow couldn't target JDK 8 anymore, which seems unlikely since various OpenJDK providers still update it (but not impossible, I suppose).

**Comment by joshtynjala** [comment]
It's probably time to consider adding a dialog to Moonshine where someone can add an arbitrary number of JDKs with a variety of versions, similar to our AS3 SDK dialog. Currently, we have the option to define JDK 8, and to define one newer JDK that meets the requirements of eclipse.jdt.ls. That was at least JDK 11, until we discovered this issue, but it's about to change to JDK 17. We should probably have the option for users to keep targeting JDK 11, I think, but any arbitrary version would be ideal.

**Comment by joshtynjala** [comment]
> I've reverted that change, added a new constant for the wrapper

@Igazine I pulled the latest from master just now, but I don't see this change yet. Are you going to commit it?

**Comment by JoelProminic** [comment]
I created a separate issue for the JDK 17 support (#1124).  This will require more testing and updates to Moonshine SDK Installer.

**Comment by joshtynjala** [comment]
>> I've reverted that change, added a new constant for the wrapper

> @Igazine I pulled the latest from master just now, but I don't see this change yet. Are you going to commit it?

@Igazine To be clear, I'm stalled until you commit this change.

**Comment by Igazine** [comment]
@joshtynjala It has nothing to do with version mismatch, that constant is just the path to the launcher, it's not used anywhere else in the code, it's totally irrelevant, so there's nothing to commit

https://github.com/Moonshine-IDE/Moonshine-IDE/blob/848e965935e407d9142636da012ff63f093a1b91/ide/MoonshineDESKTOPevolved/src/actionScripts/languageServer/JavaLanguageServerManager.as#L105

Previous value was `org.eclipse.equinox.launcher_` which file wasn't even updated with the language server update

**Comment by Igazine** [comment]
Note `moonshine-jdt` is the wrapper which prints the process id and launches the relevant eclipse launcher

**Comment by joshtynjala** [comment]
I wrote the original code, and I'm trying to explain to you what it was meant to do, and how that behavior has been broken by changing the value of this constant. The previous value of `org.eclipse.equinox.launcher_` never needs to be updated because it's only part of a larger file name. That's why the constant name includes the word `PREFIX`. The complete file name changes between eclipse.jdt.ls updates. That complete file name is used to determine if the files in _Application Support_ need to be replaced. If a file of that name exists, then Moonshine is still using the same version as it was last time. If the file doesn't exist, it means that Moonshine needs to update to a new version.

By changing this constant to `moonshine-jdt`, Moonshine can no longer detect whether it needs to replace those files in _Application Support_. Because it is checking a different file name that never changes between updates.

**Comment by joshtynjala** [comment]
To clarify, while the original issue still occurs, the revert and adding a separate constant for `moonshine-jdt` is actually a necessary part of fixing the original issue. I am waiting for you to commit that change so that I can continue with the rest of the fix.

**Comment by Igazine** [comment]
@joshtynjala I committed the necessary changes

**Comment by joshtynjala** [comment]
@Igazine Thank you!

**Comment by joshtynjala** [comment]
Alright, I just made a couple of commits that should improve things:

1. Commit 906d639e256d6883c2795a1623b0574af79f5e62 fixes the update mechanism for eclipse.jdt.ls. It turns out that the file name of the launcher JAR isn't guaranteed the change with every update to eclipse.jdt.ls. I've updated this language server probably a dozen times, and this is the first time that file kept the same name. However, there's another JAR file name that _is_ guaranteed to change, so I switched to that instead. I confirmed that the old eclipse.jdt.ls 1.9 was correctly replaced with the new eclipse.jdt.ls 1.15 on my machine.

2. In commit ce1b8d99a94af0ef2612da004cdcebf6230241be, I changed the minimum JDK major version check from 11 to 17, since that's what eclipse.jdt.ls requires now. Now, if you try to use an older JDK, instead of "exited unexpectedly", it will say that you need JDK 17 and what your current version is that's too old.

That should be all that's needed for this issue. The rest of the work may proceed in #1124.

**Comment by JoelProminic** [comment]
I updated my build, and I'm still seeing errors like this with a new Java Gradle project on macOS Monterey 12.6.1:

	: Nov 3, 2022, 5:03:58 PM Illegal character in path at index 44: file:/Users/joelanderson/Library/Application Support/com.moonshine-ide.development/Local Store/java/jdt-language-server/config_mac/org.eclipse.osgi/50/0/.cp/gradle/init/init.gradle
	: Illegal character in path at index 44: file:/Users/joelanderson/Library/Application Support/com.moonshine-ide.development/Local Store/java/jdt-language-server/config_mac/org.eclipse.osgi/50/0/.cp/gradle/init/init.gradle
	: java.net.URISyntaxException: Illegal character in path at index 44: file:/Users/joelanderson/Library/Application Support/com.moonshine-ide.development/Local Store/java/jdt-language-server/config_mac/org.eclipse.osgi/50/0/.cp/gradle/init/init.gradle
	...
	: Could not run phased build action using connection to Gradle installation '/Users/joelanderson/Downloads/MoonshineSDKs/Gradle/gradle-5.4.1-bin'.
	: org.gradle.tooling.BuildException: Could not run phased build action using connection to Gradle installation '/Users/joelanderson/Downloads/MoonshineSDKs/Gradle/gradle-5.4.1-bin'.
	...
	: Caused by: org.gradle.api.reflect.ObjectInstantiationException: Could not create an instance of type init11905159627022091243_64cgypm8s79ggrgtuonwkirwh.
	...
	: Caused by: java.lang.NoClassDefFoundError: Could not initialize class org.codehaus.groovy.runtime.InvokerHelper

Checking the init.gradle file, I don't see an obvious problem:

```
allprojects {
    project.plugins.withId("eclipse", {
        // Add prefix when project name has conflicts with root project name
        if (project != rootProject && rootProject.name.toLowerCase() == project.name.toLowerCase()) {
            eclipse.project.name = (rootProject.name + project.path).replaceAll(':', '-')
        }
    })
}
```


If I restart MoonshineDevelopment and reopen the project, I get the same error, plus another similar error:

	: Could not run phased build action using connection to Gradle installation '/Users/joelanderson/Downloads/MoonshineSDKs/Gradle/gradle-5.4.1-bin'.
	: org.gradle.tooling.BuildException: Could not run phased build action using connection to Gradle installation '/Users/joelanderson/Downloads/MoonshineSDKs/Gradle/gradle-5.4.1-bin'.
	...
	: Caused by: org.gradle.internal.exceptions.LocationAwareException: The specified initialization script '/var/folders/_g/39dkp5w56sx0qg501b5bxmcm0000gn/T/init11905159627022091243.gradle' does not exist.

**Comment by JustinProminic** [comment]
@joshtynjala I would like to understand what changed that is no part of the Moonshine package which broke the language server.  It seems like there are some external dependencies that are happening at runtime or something related to your above comment about the changes you made on eclipse.jdt.ls and the files that do and don't change.  

My concern is that I have kept the JDK stable on my machines and the language server went from working to not working one day.   Was this because I simply installed Moonshine (production 3.3.3) alongside Moonshine Development (3.3.4 build nightly) and somehow the two installations interact and override something about the language server?

I want to make sure to the greatest degree possible that Moonshine and its files are all contained in the installer as a whole.  So if we have a version release, it should continue to work for the next several years (ideally indefinitely) on an OS that is still meeting the requirements it had at the time of release.   

Please help me understand how we achieve this goal of stability on all components being contained within our distributable.   Thank you!

**Comment by joshtynjala** [comment]
@JustinProminic There is no conflict between Moonshine 3.3.3 and Moonshine 3.3.4. This issue is _not_ caused by anything that is external to the Moonshine package. The Moonshine 3.3.3 production package contains a bug. We're currently working to resolve the bug in the Moonshine Development (3.3.4 build nightly). To put it another way, the most recent Moonshine package that should work with the JDK 11 that has remained stable on your machine is Moonshine 3.3.2.

**Comment by joshtynjala** [comment]
A fresh summary of the bug:

In Moonshine 3.3.3, I updated from a version of the language server that required JDK 11 to a version that required JDK 17. I didn't realize that this JDK requirement had changed. In fact, I tested Moonshine with JDK 11 before committing, and everything seemed to be working. That's because the code I wrote to check if the included language server was updated failed.

This failure was caused by two things:

1) Tamas changed a constant referencing part of a file name. The file name changes between language server versions. I was using this constant to check to see if the language server had been updated, which causes some files to be copied to a particular location.
2) It turns out that the file name that I was checking might not actually change between versions. I should have been checking a different file name all along.

Both of those issues have been resolved now. However, because the language server still requires a newer JDK, there are a few more tasks that need to be completed before Moonshine Development is fully restored to stability. That's issue #1124.

**Comment by joshtynjala** [comment]
> Illegal character in path at index 44: file:/Users/joelanderson/Library/Application Support/com.moonshine-ide.development/Local Store/java/jdt-language-server/config_mac/org.eclipse.osgi/50/0/.cp/gradle/init/init.gradle

I saw this error once on my machine, but when I closed my project and opened it again (or maybe exited and reopened Moonshine), I never saw it again.

For reference, index 44 in the path from the error message is the space between _Application_ and _Support_ in that URL. On my machine, I recall that it was index 43 because my user name is shorter than yours by one character.

There is, in fact, a newer update to eclipse.jdt.ls that we could try, if this error message persists. Of course, I would be extra diligent about applying this update, considering what happened last time.

**Comment by rat-moonshine** [comment]
> In commit https://github.com/Moonshine-IDE/Moonshine-IDE/commit/ce1b8d99a94af0ef2612da004cdcebf6230241be, I changed the minimum JDK major version check from 11 to 17, since that's what eclipse.jdt.ls requires now. Now, if you try to use an older JDK, instead of "exited unexpectedly", it will print to the console that you need JDK 17. It will also print your current version that's too old.

Thanks for this update @joshtynjala . Last night I was talking with @JoelProminic on how to make this clear to the user that JDK17 is the latest *MUST* requirement in Moonshine, and not <17. I think the version-check and message on console will be a clear help to this.

**Comment by joshtynjala** [comment]
> java.net.URISyntaxException: Illegal character in path at index

I have confirmed that this exception was a bug in eclipse.jdt.ls 1.15.0 when certain paths contained spaces. I have upgraded to 1.16.0, which contains the fix for this issue.
